File: src/main/java/com/rtomyj/yugiohAPI/model/product/Product.java
package com.rtomyj.yugiohAPI.model.product;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import java.util.Date;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.rtomyj.yugiohAPI.controller.product.ProductController;

import com.rtomyj.yugiohAPI.helper.constants.SwaggerConstants;
import io.swagger.annotations.ApiModelProperty;
import lombok.NonNull;
import org.springframework.hateoas.RepresentationModel;

import io.swagger.annotations.ApiModel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.With;


@Data
@Builder
@With
@EqualsAndHashCode(callSuper=false)
@ToString
@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(Include.NON_EMPTY)
@ApiModel(
		description = "Information about a Yu-Gi-Oh! product such as a booster pack or tin."
)
public class Product extends RepresentationModel<Product>
{

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_ID_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private String productId;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_LOCALE_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private String productLocale;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_NAME_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private String productName;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_TYPE_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private String productType;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_SUB_TYPE_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private String productSubType;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_RELEASE_DATE_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	@JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
	private Date productReleaseDate;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_TOTAL_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private Integer productTotal;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_RARITY_STATS_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private Map<String, Integer> productRarityStats;

	@ApiModelProperty(
			value = SwaggerConstants.PRODUCT_CONTENT_DESCRIPTION
			, accessMode = ApiModelProperty.AccessMode.READ_ONLY
	)
	private List<ProductContent> productContent;

	private static final Class<ProductController> controllerClass = ProductController.class;


	private void setLink()
	{
		this.add(
			linkTo(methodOn(controllerClass).getProduct(productId, productLocale)).withSelfRel()
		);
	}


	public void setLinks()
	{
		this.setLink();
		if (this.productContent != null)	ProductContent.setLinks(this.productContent);	// set links for pack contents
	}


	public static void setLinks(@NonNull final List<Product> packs)
	{

		packs
			.stream()
			.forEach(pack -> pack.setLinks());
	}

}

File: src/main/kotlin/com/rtomyj/skc/model/Product.kt
package com.rtomyj.skc.model

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonInclude
import com.rtomyj.skc.util.constant.SwaggerConstants
import io.swagger.v3.oas.annotations.media.Schema
import java.util.*

@JsonInclude(
  JsonInclude.Include.NON_EMPTY
)
data class Product(
  @Schema(ref = "productID")
  val productId: String,
  @Schema(ref = "locale")
  val productLocale: String) {
  @Schema(ref = "productName")
  var productName: String? = null

  @Schema(ref = "productType")
  var productType: String? = null

  @Schema(ref = "productSubType")
  var productSubType: String? = null

  @Schema(
    implementation = Date::class,
    description = SwaggerConstants.PRODUCT_RELEASE_DATE_DESCRIPTION,
    pattern = "yyyy-MM-dd",
    format = "yyyy-MM-dd",
  )
  @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
  var productReleaseDate: Date? = null

  @Schema(
    implementation = Int::class,
    description = SwaggerConstants.PRODUCT_TOTAL_DESCRIPTION,
  )
  var productTotal: Int? = null

  @Schema(
    implementation = Map::class,
    description = SwaggerConstants.PRODUCT_RARITY_STATS_DESCRIPTION,
  )
  var productRarityStats: Map<String, Int>? = null

  @Schema(
    implementation = MutableList::class,
    description = SwaggerConstants.PRODUCT_CONTENT_DESCRIPTION,
  )
  var productContent: MutableList<ProductContent> = mutableListOf()
}